Adds user creation and view

shinjanxp opened the pull request:
Features include:

- Adds user model and migration
- Migrates existing user information from Availabilities, Requirements and Manufacturings to Users table with the help of a seeder.
- Creates user if necessary during ppe creation
- Sends mail to user upon successful ppe creation
- Adds multiple ppe creation support at once via ppe create form

This PR includes some features from map-ux branch as well which needed to be released before map-ux was ready to merge:

- Adds resolved and connected fields to Availabilities, Requirements and Manufacturings along with necessary modifications to Admin panel to edit them properly.
- Adds marker modifications and legends.
- Adds webhook message after ppe create
- Modifies create button
- Makes quantity optional

shinjanxp committed:
feat: Adds resolved, connected fields for all application models

shinjanxp committed:
feat: Adds beforeEdit hooks to admin to prevent sequelize errors when submitting false values to BOOLEAN fields.

shinjanxp committed:
feat: Adds purple markers for connected applications

shinjanxp committed:
fix: Adds code to api.js from index.js after refactor

shinjanxp committed:
feat: Associates application models with user

shinjanxp committed:
feat: Adds seeder which moves user info from application models

shinjanxp committed:
feat: Adds verified field to User

shinjanxp committed:
feat: Adds map legend and UI language

shinjanxp committed:
feat: Makes quantity optional

shinjanxp committed:
feat: Adds support for webhook message after application creation

Solves issue #22

shinjanxp committed:
feat: Adds Create request button in place of FAB

shinjanxp committed:
feat: Adds random token to user model

shinjanxp committed:
feat: Adds user creation to ppe/create

shinjanxp committed:
feat: Adds mailer support

Sends mail when new application is created with status view link

shinjanxp committed:
feat: Adds user route and view

shinjanxp committed:
feat: Updates thank you page with manufacturing

shinjanxp committed:
refactor: Moves ppe-create.js to javascripts folder

shinjanxp committed:
feat: Adds multiple ppe record creation to ppe-create

oyeb commented on migrations/20200414062630-augment-resolved-connected.js:
```diff
@@ -0,0 +1,79 @@
+    );
+    await queryInterface.addColumn(
+      'Availabilities',
+      'connected',
```
I'm assuming connected implies that the request (application) has been resolved by our volunteers

shinjanxp replied:
Not quite. The volunteers required 2 kinds of status

- connected: whether the application has been matched with some other application. Like a requirement being matched with a manufacturer
- resolved: the application was completely resolved, i.e. PPE was actually transferred from one party to another. Since we have no way to automatically track this, volunteers have to manually check.

oyeb commented on public/stylesheets/style.css:
```diff
@@ -92,4 +92,25 @@ a {
+  z-index: 499;
+  /* top:70px; */
+  /* left:5px; */
+  font-family: 'Times New Roman', Times, serif;
```
Seriously?

shinjanxp replied:
I'll admit we could do better here

oyeb commented on routes/ppe.js:
```diff
@@ -64,39 +72,99 @@ router.post('/',
+          password: await argon2.hash(req.body.name),
+          createdAt: new Date(),
+          updatedAt: new Date(),
+          token: cryptoRandomString({ length: 50, type: 'url-safe' })
```
If you don't mind, this could be
```
token: nanoid(50)
```
> Also in `move-existing-user-data.js`

shinjanxp replied:
Alright, sure.

oyeb commented on routes/ppe.js:
```diff
@@ -64,39 +72,99 @@ router.post('/',
+        // Preview URL: https://ethereal.email/message/WaQKMgKddxQDoou...  
+      });
+
+      console.log("USER: " + user.id)
```
Unnecessary, IMO the app should save the token in a cookie. Perhaps, that's been planned?

oyeb commented on routes/ppe.js:
```diff
@@ -64,39 +72,99 @@ router.post('/',
+          r.remarks = req.body.remarks;
+        }
+        const manufacturing = await models.Manufacturing.bulkCreate(records);
+        return res.render('ppe-thanks', { forId: manufacturing.id, forType: 'Manufacturing', statusLink: statusLink });
```
All 3 renders could be refactored out of the conditionals

shinjanxp replied:
yes right. We had a different render for manufacturing till now, so it seemed a good idea to have them inside separate conditionals, but now we could move them out into one. Good catch

oyeb commented on views/ppe-thanks.ejs:
```diff
@@ -1,15 +1,21 @@
+                Once we match someone with you, we will send you an email notification or contact you directly
+                in case of urgency.
+            </p>
+            <p class="lead">If you wish to receive notifications in addition to emails, please click the Subscribe button below.</p>
```
On L28 there's an unnecessary GET of `/config.js`

shinjanxp replied:
No it's being used by the `subscribeUserToPush` function on L96

oyeb reviewed (approved):
I like the idea of user tokens. We're so close to automatically logging returning users in when they return!
Nice work!

shinjanxp committed:
chore: Improves configuration and README

shinjanxp committed:
perf: Uses nanoid and improves ppe-create code flow

ref: https://github.com/covid-19-ppe-tracker/covid-19-ppe-tracker/pull/23#discussion_r416052620
https://github.com/covid-19-ppe-tracker/covid-19-ppe-tracker/pull/23#discussion_r416050511

shinjanxp committed:
feat: Improves UI of list display and changes create button to add
